File: TicketBookingsAppFE/src/app/models/Event.ts
export interface Event {
  eventID: string; // Guid in the backend is mapped to a string in TypeScript
  name: string;
  dateAndTime: string; // DateTimeOffset maps to string (ISO 8601 format)
  venue: Venue; // Reference the VenueDTO structure
  eventDescription?: string; // Optional
  categories: string[]; // string enum
  ticketTypes: TicketType[]; // List of TicketTypeDTO
  images: EventImage[]; // list of EventImageDTO
  organizer: Organizer; // OrganizerDTO
  availableTickets: number; // This is a calculated field
  totalTicketPrice: number; // This is a calculated field
}

// Venue DTO structure
export interface Venue {
  name: string;
  address: string;
  state: string;
  capacity: number;
}

// TicketType DTO structure
export interface TicketType {
  type: string;
  price: number;
  quantityAvailable: number;
  ticketTypeID: string;
}

// EventImage DTO structure
export interface EventImage {
  imageUrl: string;
}

// Organizer DTO structure
export interface Organizer {
  name: string;
  contactEmail: string;
  phoneNumber: string;
}

export enum EventCategory {
  Music = 'Music',
  Sports = 'Sports',
  Technology = 'Technology',
  Education = 'Education',
  Business = 'Business',
  Arts = 'Arts',
  Health = 'Health',
  Entertainment = 'Entertainment',
  FoodAndDrink = 'FoodAndDrink',
  Fashion = 'Fashion',
  Charity = 'Charity',
  Comedy = 'Comedy',
  FilmAndTheater = 'FilmAndTheater',
  Literature = 'Literature',
  Networking = 'Networking',
  Science = 'Science',
  TravelAndAdventure = 'TravelAndAdventure',
  Workshop = 'Workshop',
  Conference = 'Conference',
  Festival = 'Festival',
  Religious = 'Religious',
  Fitness = 'Fitness',
  Family = 'Family',
  Gaming = 'Gaming',
  History = 'History',
  Politics = 'Politics',
  Spirituality = 'Spirituality',
  Environment = 'Environment',
  Photography = 'Photography',
  PetsAndAnimals = 'PetsAndAnimals',
}

//` Function to map the API response to Event model
export function mapToEvent(item: any): Event {
  return {
    eventID: item.eventID,
    name: item.name,
    dateAndTime: item.dateAndTime,
    venue: {
      name: item.venue.name,
      address: item.venue.address,
      state: item.venue.state,
      capacity: item.venue.capacity,
    },
    eventDescription: item.eventDescription || '',
    categories: item.categories.$values ? item.categories.$values : [], // Assuming this is an array of enums or strings
    ticketTypes: item.ticketTypes.$values.map((tt: any) => ({
      type: tt.type,
      price: tt.price,
      quantityAvailable: tt.quantityAvailable,
      ticketTypeID: tt.ticketTypeID,
    })),
    images: item.images.$values.map((img: any) => ({
      imageUrl: img.imageUrl,
    })),
    organizer: {
      name: item.organizer.name,
      contactEmail: item.organizer.contactEmail,
      phoneNumber: item.organizer.phoneNumber,
    },
    availableTickets: item.availableTickets,
    totalTicketPrice: item.totalTicketPrice,
  };
}

File: TicketBookingsAppAPI/Models/Domain/Event.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace TicketBookingsAppAPI.Models.Domain
{
    public class Event
    {
        [Key]  // Primary Key
        public Guid EventID { get; set; }

        [Required]
        [MaxLength(100)]  // Restrict the name to 100 characters
        public string Name { get; set; }

        [Required]
        public DateTimeOffset DateAndTime { get; set; }

        // Embedded Venue information
        [Required]
        public Venue Venue { get; set; }

        // Collection of TicketTypes with FK to Event
        public List<TicketType> TicketTypes { get; set; } = new List<TicketType>();

        [MaxLength(2000)]  // Restrict event description to 2000 characters
        public string? EventDescription { get; set; }

        // Categories (multi-select)
        [Required]
        public List<EventCategory> Categories { get; set; } = new List<EventCategory>();

        // Calculated field for available tickets (not mapped to DB)
        [NotMapped]  // Do not map this to the database
        public int AvailableTickets
        {
            get
            {
                return TicketTypes.Sum(ticket => ticket.QuantityAvailable);
            }
        }

        // Event Images with FK to Event
        public List<EventImage> Images { get; set; } = new List<EventImage>();

        // Embedded Organizer information
        [Required]
        public Organizer Organizer { get; set; }

        //public int TicketPrice { get; internal set; }
    }

    // Venue Class for detailed venue information (embedded)
    public class Venue
    {
        [Required]
        [MaxLength(255)]  // Venue name length
        public string Name { get; set; }

        [Required]
        [MaxLength(500)]  // Venue address length
        public string Address { get; set; }

        [Required]
        [MaxLength(100)]  // State length
        public string State { get; set; }

        [Required]
        public int Capacity { get; set; }
    }    

    // Organizer Class (embedded)
    public class Organizer
    {
        [Required]
        [MaxLength(255)]  // Organizer name length
        public string Name { get; set; }

        [Required]
        [EmailAddress]  // Enforce valid email
        public string ContactEmail { get; set; }

        [Phone]  // Enforce valid phone number
        public string PhoneNumber { get; set; }
    }

    // Enum for Event Categories
    public enum EventCategory
    {
        Music,
        Sports,
        Technology,
        Education,
        Business,
        Arts,
        Health,
        Entertainment,
        FoodAndDrink,
        Fashion,
        Charity,
        Comedy,
        FilmAndTheater,
        Literature,
        Networking,
        Science,
        TravelAndAdventure,
        Workshop,
        Conference,
        Festival,
        Religious,
        Fitness,
        Family,
        Gaming,
        History,
        Politics,
        Spirituality,
        Environment,
        Photography,
        PetsAndAnimals
    }

}
